src/avx512-64bit-keyvaluesort.hpp

```diff
@@ -0,0 +1,1276 @@
+                                          left + 3 * size,
+                                          left + 4 * size,
+                                          left + 5 * size,
+                                          left + 6 * size,
+                                          left + 7 * size,
+                                          left + 8 * size);
+    zmm_t key_vec = vtype::template i64gather<sizeof(type_t)>(rand_index, keys);
+
+    index_t index_vec;
+    zmm_t sort;
+    index_vec = zmm_kv_vector<uint64_t>::template i64gather<sizeof(uint64_t)>(
+            rand_index, indexes);
+    sort = sort_zmm_64bit<vtype>(key_vec, index_vec);
+    // pivot will never be a nan, since there are no nan's!
+
+    return ((type_t *)&sort)[4];
+}
+
+template <typename vtype, typename type_t>
+inline void heapify(type_t *keys, uint64_t *indexes, int64_t idx, int64_t size)
+{
+    int64_t i = idx;
+    while (true) {
+        int64_t j = 2 * i + 1;
+        if (j >= size || j < 0) { break; }
+        int k = j + 1;
+        if (k < size && keys[j] < keys[k]) { j = k; }
+        if (keys[j] < keys[i]) { break; }
+        std::swap(keys[i], keys[j]);
+        std::swap(indexes[i], indexes[j]);
+        i = j;
+    }
+}
+template <typename vtype, typename type_t>
+inline void heap_sort(type_t *keys, uint64_t *indexes, int64_t size)
+{
+    for (int64_t i = size / 2 - 1; i >= 0; i--) {
+        heapify<vtype>(keys, indexes, i, size);
+    }
+    for (int64_t i = size - 1; i > 0; i--) {
+        std::swap(keys[0], keys[i]);
+        std::swap(indexes[0], indexes[i]);
+        heapify<vtype>(keys, indexes, 0, i);
+    }
+}
+
+template <typename vtype, typename type_t>
+inline void qsort_64bit_(type_t *keys,
+                         uint64_t *indexes,
+                         int64_t left,
+                         int64_t right,
+                         int64_t max_iters)
+{
+    /*
+     * Resort to std::sort if quicksort isnt making any progress
+     */
+    if (max_iters <= 0) {
+        heap_sort<vtype>(keys + left, indexes + left, right - left + 1);
+        return;
+    }
+    /*
+     * Base case: use bitonic networks to sort arrays <= 128
+     */
+    if (right + 1 - left <= 128) {
+
+        sort_128_64bit<vtype>(
+                keys + left, indexes + left, (int32_t)(right + 1 - left));
+        return;
+    }
+
+    type_t pivot = get_pivot_64bit<vtype>(keys, indexes, left, right);
+    type_t smallest = vtype::type_max();
+    type_t biggest = vtype::type_min();
+    int64_t pivot_index = partition_avx512<vtype>(
+            keys, indexes, left, right + 1, pivot, &smallest, &biggest);
+    if (pivot != smallest)
+        qsort_64bit_<vtype>(
+                keys, indexes, left, pivot_index - 1, max_iters - 1);
+    if (pivot != biggest)
+        qsort_64bit_<vtype>(keys, indexes, pivot_index, right, max_iters - 1);
+}
+
+X86_SIMD_SORT_INLINE int64_t replace_nan_with_inf_kv(double *arr,
```

Comment:
This is the same function as in `avx512-64bit-qsort.hpp`. You could move this to the common header file as well and no need to re-write it.